File: frontend/src/utils/urls.js
export function publicProductUrl(url) {
  if (!url) return "";
  try {
    const parsed = new URL(url);
    if (["example.com", "www.example.com"].includes(parsed.hostname)) return "";
    if (!["http:", "https:"].includes(parsed.protocol)) return "";
    return parsed.toString();
  } catch {
    return "";
  }
}

export function openPublicProductUrl(url) {
  const publicUrl = publicProductUrl(url);
  if (!publicUrl) return false;
  window.open(publicUrl, "_blank", "noopener,noreferrer");
  return true;
}

export const PRODUCT_LINK_UNAVAILABLE_TEXT = "公开 Demo 暂无真实下单链接";

File: backend/app/utils/urls.py
from urllib.parse import urlparse


PLACEHOLDER_HOSTS = {"example.com", "www.example.com"}


def public_product_url(url: str | None) -> str | None:
    if not url:
        return None
    parsed = urlparse(str(url).strip())
    if parsed.hostname in PLACEHOLDER_HOSTS:
        return None
    if parsed.scheme not in {"http", "https"}:
        return None
    return str(url).strip()
